quodlibet/quodlibet/util/collection.py

```diff
@@ -595,38 +612,148 @@ def from_songs(cls, dir_, songs, library=None):
+            node = tree.find("title")
+            if self.name != node.text:
+                print_w("Found name \"%s\" instead of \"%s\" for %s"
+                        % (node.text, self.name, self.path))
+            for node in tree.iterfind('.//track'):
+                location = node.find('location')
+                path = location.text.replace('\n', '').replace('\r', '')
+                if path in library:
+                    self._list.append(library[path])
+                elif library and library.masked(path):
+                    self._list.append(path)
+                else:
+                    # TODO: handle missing playlist items (#3105, #729, #3131)
+                    print_w("Couldn't find %s in playlist at %s, "
+                            "but perhaps its metadata will help: %s"
+                            % (path, self.path, ET.tostring(node)))
+                    self._list.append(path)
+                    library.mask(path)
+        except ET.ParseError as e:
+            print_w("Couldn't load %s (%s)" % (self.path, e))
+
+    @classmethod
+    def filename_for(cls, name: str) -> str:
+        # Manually do *minimal* escaping, to allow near-readable filenames
+        for bad, good in cls._SAFER.items():
+            name = name.replace(bad, quote(good, safe=''))
+        return path2fsn("%s.%s" % (limit_path(name), cls.EXT))
+
+    @classmethod
+    def name_for(cls, filename: fsnative) -> str:
+        filename, ext = splitext(unescape_filename(filename))
+        if not ext or ext.lower() != (".%s" % cls.EXT):
+            raise TypeError("XSPFs should end in '.%s', not '%s'"
+                            % (cls.EXT, ext))
+        return filename
+
+    def write(self):
+        track_list = Element("trackList")
+        for song in self._list:
+            if isinstance(song, str):
+                track = {"location": song}
+            else:
+                creator = self.CREATOR_PATTERN.format(song)
+                track = {
+                    "location": song("~filename"),
+                    "title": song("title"),
+                    "creator": creator,
+                    "album": song("album"),
+                    "trackNum": song("~#track"),
+                    "duration": int(song("~#length") * 1000.)
+                }
+            track_list.append(self.element_from("track", track))
+        playlist = Element("playlist", attrib={"version": "1"})
+        playlist.append(self.text_el("title", self.name))
+        playlist.append(self.text_el("date", datetime.now().isoformat()))
+        playlist.append(track_list)
+        tree = ElementTree(playlist)
+        ET.register_namespace('', XSPF_NS)
+        tree.write(self.path, encoding="UTF-8", xml_declaration=True)
+        if self._last_fn != self.path:
+            self._delete_file(self._last_fn)
+            self._last_fn = self.path
+
+    @classmethod
+    def text_el(cls, name: str, value: any) -> Element:
```

Comment:
minor thing, but I think for consistency this should also have "element" in the name.

Reply:
Good shout, I'll update that